### PR TITLE
systemboot: control debug level via VPD variable

### PR DESCRIPTION
**Pull request by tfg13**
Continuing https://github.com/u-root/u-root/pull/1869 here.

Accepting coreboot loglevels now, (but still mapping them internally to quiet/verbose as that is all we have for now)

Testing done, I built a new image, set `systemboot_log_level=crit` and flashed it to a real devices.

Log lines from systemboot before: 219
Log lines from systemboot now: 117

DHCP debug dumps are quiet now, but it is still visible what is going one. There is still too much output from the disk mounting, but way less than before.

### CODE REVIEW

**Comment by tfg13** [comment]
Updated tests, see above

File: cmds/boot/systemboot/main.go [hunk 1]
```diff
@@ -208,14 +242,65 @@ func addSEL(sequence string) {
...
+// the VPD variable "systemboot_log_level".
+// Valid values are coreboot loglevels https://review.coreboot.org/cgit/coreboot.git/tree/src/commonlib/include/commonlib/loglevel.h,
+// either as integer (1, 2) or string ("debug").
+// If the VPD variable is missing or it is set to an invalid value, it will use the default.
+func getDebugEnabled() bool {
+	if isFlagPassed("q") {
+		return !*doQuiet
+	}
+
+	// -q was not passet, so `doQuiet` contains the default value
```
Comment (rojer9-fb) [inline cmds/boot/systemboot/main.go:257]: typo: passed

File: cmds/boot/systemboot/main.go [hunk 2]
```diff
@@ -208,14 +242,65 @@ func addSEL(sequence string) {
...
+		log.Printf("Warning: failed to read read-write VPD variable '%s', will try the read-only one. Error was: %v", vpdSystembootLogLevel, err)
+		v, err = vpd.Get(vpdSystembootLogLevel, true)
+		if err != nil {
+			// TODO do not print warning if file is not found
+			log.Printf("Warning: failed to read read-only VPD variable '%s', will use the default value. Error was: %v", vpdSystembootLogLevel, err)
+			return defaultDebugEnabled
+		}
+	}
+	level := strings.ToLower(strings.TrimSpace(string(v)))
+	// allow integer and translate to string level
```
Comment (rojer9-fb) [inline cmds/boot/systemboot/main.go:272]: good idea but i'd move it to the "default" branch of the switch statement below

File: cmds/boot/systemboot/main.go [hunk 3]
```diff
@@ -208,14 +242,65 @@ func addSEL(sequence string) {
...
+	if isFlagPassed("q") {
+		return !*doQuiet
+	}
+
+	// -q was not passet, so `doQuiet` contains the default value
+	defaultDebugEnabled := !*doQuiet
+	// check for the VPD variable "systemboot_log_level". First the read-write, then the read-only
+	v, err := vpd.Get(vpdSystembootLogLevel, false)
+	if err != nil {
+		// TODO do not print warning if file is not found
```
Comment (rojer9-fb) [inline cmds/boot/systemboot/main.go:262]: um... why not DO it now?

File: cmds/boot/systemboot/main.go (hunk 2 above)
Comment (tfg13) [reply cmds/boot/systemboot/main.go:272]: not sure I follow, then I need to do the mapping of levels to debug on/off again?

File: cmds/boot/systemboot/main.go (hunk 3 above)
Comment (tfg13) [reply cmds/boot/systemboot/main.go:262]: Ha I copied this from the original PR, not sure there is even a separate error for this, checking

File: cmds/boot/systemboot/main.go (hunk 2 above)
Comment (rojer9-fb) [reply cmds/boot/systemboot/main.go:272]: strconv stuff should be attempted only if the string didn't match any of the known levels, i.e. in the default branch of the switch statement below.

Comment (rojer9-fb) [reply cmds/boot/systemboot/main.go:272]: oh, sorry, i misread. the logic is backwards.
ok, ignore this.

**Review by rojer9-fb — APPROVED** [review]
LGTM. if possible, get rid of TODOs if too much work - ok to check in as is.

File: cmds/boot/systemboot/main.go [hunk 4]
```diff
@@ -208,14 +242,65 @@ func addSEL(sequence string) {
...
+		log.Printf("Warning: failed to read read-write VPD variable '%s', will try the read-only one. Error was: %v", vpdSystembootLogLevel, err)
+		v, err = vpd.Get(vpdSystembootLogLevel, true)
+		if err != nil {
+			// TODO do not print warning if file is not found
+			log.Printf("Warning: failed to read read-only VPD variable '%s', will use the default value. Error was: %v", vpdSystembootLogLevel, err)
+			return defaultDebugEnabled
+		}
+	}
+	level := strings.ToLower(strings.TrimSpace(string(v)))
```
Comment (rminnich) [inline cmds/boot/systemboot/main.go:271]: I think you can skip a lot of this. 
I don't see the need to convert to a string (which you don't use?) when you can catch the case below?
switch (strings.TrimSpace(string(v))) {
case "whatrever"
case "0":
case "1":
etc.
or even 
case "0", "1", "2", "3": 
return false
case "4", etc.
default:
return defaultDebugEnabled

and so on. 
lines 271-280 seem unnecessary

**Review by rminnich — CHANGES_REQUESTED** [review]
overall, looks nice, please see comments

File: cmds/boot/systemboot/main.go (hunk 4 above)
Comment (tfg13) [reply cmds/boot/systemboot/main.go:271]: Hmm not sure I follow. I am using the string levels (in the switch).
I just translate integer levels (if any) to the equivalent string to make the switch afterwards more readable.

File: cmds/boot/systemboot/main.go [hunk 5]
```diff
@@ -6,26 +6,60 @@ package main
...
+	"warning",
+	"notice",
+	"info",
+	"debug",
+	"spew",
+	"never",
+}
+
+// isFlagPassed checks whether a flag was explicitly passed on the command line. Thank you StackOverflow.
+func isFlagPassed(name string) bool {
```
[the remark this answers is not in the document]
Comment (tfg13) [reply cmds/boot/systemboot/main.go:53]: I tried, but it returns non-nil even if not set, so I have no way to tell if it is not set or default: https://play.golang.com/p/NmwWaWkRb3u

File: cmds/boot/systemboot/main.go (hunk 3 above)
Comment (tfg13) [reply cmds/boot/systemboot/main.go:262]: Hmm `vpd` itself doesn't guarantee any specific errors. I see it just uses a normal file open internally, so I probably could get it from there somehow, but not sure I want to depend on it. Also this is a pain to test again :/

**Comment by tfg13** [comment]
Fixed typos, reworded flag help text, moved warning level to be quiet as suggested by @johnnylinwiwynn

File: cmds/boot/systemboot/main.go (hunk 4 above)
Comment (tfg13) [reply cmds/boot/systemboot/main.go:271]: Ah I got it now, changing

File: cmds/boot/systemboot/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (rminnich) [reply cmds/boot/systemboot/main.go:37]: Ah. You can if you want see the example of creating a type which implements Set and String(); this lets you call arbitrary code when a flag is set. But I guess this works as well.

**Review by rminnich — APPROVED** [review]

**Pull request merged**